File: Data Structures/python/union_find.py
class Unionfind:
    def __init__(self, s) -> None:
        self.__size=s
        self.__id=[x for x in range (0, s)]
        self.__sz=[1]*s
        self.__numComponents=s
    
    def find(self, p):
        root=p
        while(self.__id[root]!=root):  root=self.__id[root]

        if(root==p): return root
        
        while(p!=root):
            next=self.__id[p]
            self.__id[p]=root
            p=next
        
        return root
    
    def connected(self, p, q):
        return self.find(p)==self.find(q)
    
    def componentSize(self, p):
        return self.__sz[self.find(p)]
    
    def size(self):
        return self.__size
    
    def components(self):
        return self.__numComponents
    
    def unify(self, p, q):
        if(self.connected(p, q)): return
        root1=self.find(p)
        root2=self.find(q)
        if(self.__sz[root1]<self.__sz[root2]):
            self.__sz[root2]+=self.__sz[root1]
            self.__id[root1]=root2
            self.__sz[root1]=0
        else:
            self.__sz[root1]+=self.__sz[root2]
            self.__id[root2]=root1
            self.__sz[root2]=0
    
s=Unionfind(10)
for i in range (3, 10, 2):
    s.unify(i-2, i)
for i in range (0, 10):
    print( i, "th component size: ", s.componentSize(i))

File: Data Structures/cpp/union_find.cpp
#include<iostream>
#include<vector>
using namespace std;

class UnionFind{
    private:
        int size;
        vector<int> sz;
        vector<int> id;
        int numComponents=0;
    
    public:
        UnionFind(int s){
            this->size=numComponents=s;
            sz.resize(size);
            id.resize(size);
            for(int i=0; i<size; i++){
                sz[i]=1;
                id[i]=i;
            }
        }

        int find(int p){
            int root =p;
            while(id[root]!=root) root=id[root];

            //compressing the path
            while(p!=root){
                int next=id[p];
                id[p]=root;
                p=next;
            }

            return root;
        }
        
        bool connected(int p, int q){
            return find(p)==find(q);
        }

        int componentSize(int p){
            return sz[find(p)];
        }

        int len(){
            return this->size;
        }

        int components(){
            return numComponents;
        }

        void unify(int p, int q){

            if(connected(p, q)) return;

            int root1=find(p), root2=find(q);
            if(sz[root1]<sz[root2]){
                sz[root2]+=sz[root1];
                id[root1]=root2;
                sz[root1]=0;
            }
            else{
                sz[root1]+=sz[root2];
                id[root2]=root1;
                sz[root2]=0;
            }
            numComponents--;
        }

};

int main(){
    UnionFind s(10);
    cout<<s.components()<<endl;
    for(int i=3; i<10; i+=2){
        s.unify(i-2, i);
        cout<<i<<"st unify: "<<s.components()<<endl;
    }
    for(int i=0; i<10; i++)
        cout<<i<<" component size: "<< s.componentSize(i)<<endl;
    return 0;
}